src/testset/views.py

```diff
@@ -21,18 +29,118 @@ def get_context_data(self, *, object_list=None, **kwargs):
-    context_object_name = 'result_list'
+    context_object_name = 'user_list'
 
     paginate_by = 5
 
-    def get_queryset(self):
-        qs = super().get_queryset().order_by('-avg_score')
-        qs.select_related('user')
-        return qs
+    # def get_queryset(self):
+    #     qs = super().get_queryset().order_by('-avr_score')
+    #     qs.select_related('user')
+    #     return qs
 
     def get_context_data(self, *, object_list=None, **kwargs):
         context = super().get_context_data(object_list=None, **kwargs)
         context['title'] = 'LeaderBoard'
         return context
+
+
+class TestRunView(View):
+
+    PREFIX = 'answer_'
+    variants_count = []
+
+    def get(self, request, pk, seq_nr):
+        print(self.variants_count)
+
+        question = Question.objects.filter(test__id=pk, number=seq_nr).first()
+
+        variants = [
+            variant.text
+            for variant in question.variants.all()
+        ]
+        return render(request, 'testset/test_run.html', {'question': question,
+                                                         'answers': variants,
+                                                         'prefix': self.PREFIX})
+
+    def post(self, request, pk, seq_nr):
+
+        test = Test.objects.get(pk=pk)
+        question = Question.objects.filter(test__id=pk, number=seq_nr).first()
+
+        variants = Variant.objects.filter(
+            question=question
+        ).all()
+
+        choices = {
+            k.replace(self.PREFIX, ''): True
+            for k in request.POST if k.startswith(self.PREFIX)
+        }
+        self.variants_count.append(variants.count())
+
+        if not choices:
+            messages.error(self.request, extra_tags='danger', message="ERROR: You should select at least 1 answer!")
+            return redirect(reverse('testset:run_step', kwargs={'pk': pk, 'seq_nr': seq_nr}))
+
+        current_test_result = TestResult.objects.filter(
+            test=test,
+            user=request.user,
+            is_completed=False).last()
+
+        for idx, variant in enumerate(variants, 1):
+            value = choices.get(str(idx), False)
+            TestResultDetail.objects.create(
+                test_result=current_test_result,
+                question=question,
+                variant=variant,
+                is_correct=(value == variant.is_correct)
+            )
+
+        if question.number < test.questions_count():
+            return redirect(reverse('testset:run_step', kwargs={'pk': pk, 'seq_nr': seq_nr + 1}))
+        else:
+            qs = current_test_result.test_result_details.values('question').filter(is_correct=True).\
+                annotate(Count('is_correct'))
+
+            is_correct = 0
+            for qs_item in zip(qs, self.variants_count):
+                if qs_item[0].get('is_correct__count') == qs_item[1]:
+                    is_correct += 1
+
+            self.variants_count.clear()
```

Comment:
As a part of my another comment, consider placing this into model. Actually, this is already done as a part of `update_score` method, you just need to create yet another representation of this data. Something like this, for example:
```
class TestResult(models.Model):
    ...
    def correct_answers_count(self):
        qs = self.details.values('question').annotate(
            num_answers=Count('question'),
            points=Sum('is_correct')
        )
        return sum(e['num_answers'] == int(e['points']) for e in qs)

    def score_info(self):
        num_questions = self.test.questions_count()
        num_answers = self.correct_answers_count()
        return f'{self.num_answers} of {num_questions} ({(self.num_answers/num_questions)*100:.2f}%)'
```

Reply:
Fixed.